middleware/gzip/filter.go

```diff
@@ -72,7 +84,7 @@ func DefaultMIMEFilter() MIMEFilter {
 // matches any of the registered ones. It returns true if
 // found and false otherwise.
 func (m MIMEFilter) ShouldCompress(r *http.Request) bool {
-	return m.Types.Contains(r.Header.Get("Content-Type"))
+	return m.Types.Contains(r.Header.Get("Accept"))
```

Comment:
You're almost there. Make this method check accept first and use MIME types later and we're good to go.
```
func (m MIMEFilter) ShouldCompress(r *http.Request) bool {
	ext := filepath.Ext(r.URL.Path)
	if r.Header.Get("Accept") != "" {
		return m.Types.Contains(r.Header.Get("Accept"))
	}
	return m.Types.Contains(mime.TypeByExtension(ext))
}
```